sockeye/transformer.py

```diff
@@ -0,0 +1,298 @@
+                 num_hidden: int,
+                 num_model: int,
+                 dropout: float,
+                 prefix: str) -> None:
+        self.num_hidden = num_hidden
+        self.num_model = num_model
+        self.dropout = dropout
+        self.prefix = prefix
+        self.w_i2h = mx.sym.Variable('%si2h_weight' % prefix)
+        self.b_i2h = mx.sym.Variable('%si2h_bias' % prefix)
+        self.w_h2o = mx.sym.Variable('%sh2o_weight' % prefix)
+        self.b_h2o = mx.sym.Variable('%sh2o_bias' % prefix)
+
+    def __call__(self, x, length) -> mx.sym.Symbol:
+        """
+        Position-wise feed-forward network with ReLU activation.
+
+        :param x: Symbol of shape (batch_size, seq_len, num_hidden)
+        :param length: sequence length
+        :return: Symbol of shape (batch_size, seq_len, num_hidden)
+        """
+        # TODO: use a convolution?
+        x = mx.sym.reshape(x, shape=(-3, -1))
+        h = mx.sym.FullyConnected(data=x, num_hidden=self.num_hidden, weight=self.w_i2h, bias=self.b_i2h)
+        h = mx.sym.Activation(h, act_type="relu")
+        if self.dropout > 0.0:
+            h = mx.sym.Dropout(h, p=self.dropout)
+        y = mx.sym.FullyConnected(data=h, num_hidden=self.num_model, weight=self.w_h2o, bias=self.b_h2o)
+        y = mx.sym.reshape(y, shape=(-1, length, self.num_model))
+        return y
+
+
+def get_autoregressive_bias(max_length: int, name: str) -> mx.sym.Symbol:
+    """
+    Returns bias/mask to ensure position i can only attend to positions <i.
+
+    :param max_length: Sequence length.
+    :param name: Name of symbol.
+    :return: Bias symbol of shape (1, max_length, max_length).
+    """
+    return mx.sym.BlockGrad(mx.symbol.Custom(length=max_length,
+                                             name=name,
+                                             op_type='auto_regressive_bias'))
+
+
+class AutoRegressiveBias(mx.operator.CustomOp):
+    """
+    Returns a symbol of shape (1, length, length) with cells above the main diagonal
+    set to a large negative value, e.g.
+    length=4
+
+    0 1 1 1
+    0 0 1 1   * -99999
+    0 0 0 1
+    0 0 0 0
+    """
+
+    def __init__(self, length: int):
+        super().__init__()
+        self.bias = self.get_bias(length)
+
+    @staticmethod
+    def get_bias(length: int):
+        # matrix with lower triangle and main diagonal set to 0, upper triangle set to 1
+        upper_triangle = np.triu(np.ones((length, length)), k=1)
+        # (1, length, length)
+        bias = -99999999. * np.reshape(upper_triangle, (1, length, length))
+        return mx.nd.array(bias)
+
+    def forward(self, is_train, req, in_data, out_data, aux):
+        self.assign(out_data[0], req[0], self.bias)
+
+    def backward(self, req, out_grad, in_data, out_data, in_grad, aux):
+        pass
+
+
+@mx.operator.register("auto_regressive_bias")
+class AutoRegressiveBiasProp(mx.operator.CustomOpProp):
+
+    def __init__(self, length: str):
```

Comment:
yes, thats one of the quite obscure limitations of passing arguments to python custom ops.